src/tiled/wangbrush.cpp

```diff
@@ -0,0 +1,453 @@
+#include "brushitem.h"
+#include "containerhelpers.h"
+#include "mapdocument.h"
+#include "maprenderer.h"
+#include "painttilelayer.h"
+#include "tilelayer.h"
+#include "wangset.h"
+
+#include <cmath>
+
+using namespace Tiled;
+using namespace Internal;
+
+//value between 0 and 0.5 to control the dead zone with edge mode.
+static const float MIDDLE_DEAD_ZONE = 0.25f;
+static const float EDGE_DEAD_ZONE = 0.2f;
+
+WangBrush::WangBrush(QObject *parent)
+    : AbstractTileTool(tr("Wang Brush"),
+                      QIcon(QLatin1String(
+                                ":images/24x24/wangtile-edit.png")),
+                      QKeySequence(tr("G")),
+                      parent)
+    , mEdgeDir(0)
+    , mIsActive(false)
+    , mWangSet(nullptr)
+    , mCurrentColor(0)
+    , mBrushMode(Idle)
+    , mIsTileMode(false)
+    , mBrushBehavior(Free)
+{
+}
+
+WangBrush::~WangBrush()
+{
+}
+
+void WangBrush::activate(MapScene *scene)
+{
+    mIsActive = true;
+    AbstractTileTool::activate(scene);
+}
+
+void WangBrush::deactivate(MapScene *scene)
+{
+    mIsActive = false;
+    AbstractTileTool::deactivate(scene);
+}
+
+void WangBrush::mousePressed(QGraphicsSceneMouseEvent *event)
+{
+    if (mBrushMode == Idle || !brushItem()->isVisible())
+        return;
+
+    if (event->button() == Qt::LeftButton) {
+        switch (mBrushBehavior) {
+        case Free:
+            beginPaint();
+            break;
+        default:
+            break;
+        }
+    }
+}
+
+void WangBrush::mouseReleased(QGraphicsSceneMouseEvent *event)
+{
+    switch (mBrushBehavior) {
+    case Paint:
+        if (event->button() == Qt::LeftButton)
+            mBrushBehavior = Free;
+    default:
+        break;
+    }
+}
+
+void WangBrush::modifiersChanged(Qt::KeyboardModifiers modifiers)
+{
+    if ((modifiers & Qt::ControlModifier) != mIsTileMode) {
+        stateChanged();
+        mIsTileMode = modifiers & Qt::ControlModifier;
+    }
+}
+
+void WangBrush::languageChanged()
+{
+    setName(tr("Wang Brush"));
+    setShortcut(QKeySequence(tr("G")));
+}
+
+void WangBrush::setEdgeColor(int color)
+{
+    mCurrentColor = color;
+    mBrushMode = PaintEdge;
+}
+
+void WangBrush::setCornerColor(int color)
+{
+    mCurrentColor = color;
+    mBrushMode = PaintVertex;
+}
+
+void WangBrush::mouseMoved(const QPointF &pos, Qt::KeyboardModifiers modifiers)
+{
+    if (mBrushMode == Idle)
+        return;
+
+    if (mIsTileMode) {
+        AbstractTileTool::mouseMoved(pos, modifiers);
+        return;
+    }
+
+    QPointF offsetPos = pos;
+    if (Layer *layer = currentLayer()) {
+        offsetPos -= layer->totalOffset();
+        brushItem()->setLayerOffset(layer->totalOffset());
+    }
+
+    const MapRenderer *renderer = mapDocument()->renderer();
+    const QPointF tilePosF = renderer->screenToTileCoords(offsetPos);
+    QPoint tilePos;
+
+    if (mBrushMode == PaintVertex) {
+        tilePos = tilePosF.toPoint();
+        if (tilePos != mPaintPoint) {
```

Comment:
I was using that at first, but I think the behavior I wanted needed something slightly different from PaintVertex. In any case Ill take another look over it and make sure it couldn't be made better.